crates/ra_lsp_server/src/world.rs

```diff
@@ -63,7 +63,7 @@ pub struct WorldSnapshot {
     pub workspaces: Arc<Vec<ProjectWorkspace>>,
     pub analysis: Analysis,
     pub latest_requests: Arc<RwLock<LatestRequests>>,
-    pub check_watcher: CheckState,
+    pub diagnostics: DiagnosticCollection,
```

Comment:
We should remove this field and instead add `Diagnostics(...)` variant to the `Task` enum. Then, then `handle_diagnostics` function won't be able to directly send `PublishiDiagnosticsParams` to the client. Instead, it will just publish the diagnostics as a task.

The main loop then will be able to publish the current set of diagnostics.

Reply:
> Are we supposed to schedule an update on the thread pool on each file change, and then send an event to the main loop with the result if we manage to compute diagnostics without being canceled?

In other words, yes. I am not sure if this is the best solution, but it seems a nice one

Reply:
Okay, I get what you were going for. We still need some state though, right? To be able to properly clear diagnostics we'll still need to know what files we have diagnostics for. Although it could probably just be a list of `FileId`s/`Uri`s

Reply:
>We still need some state though, right? 

Right, but the state will only be stored in a World, not in a snapshot. 

>we'll still need to know what files we have diagnostics for. 

If we define `DiagnosticsCollection` as `struct DC { check_diagostics: HashMap<FileId, Diag>, ra_diagnostics: HashMap<FileId, Diag>}`, than this is `diagnostics_collection.check_diagnostics.key()`.

Reply:
Right, that makes sense, thanks for the feedback, will get to coding in a bit :)

Reply:
We might actually need a copy of the diagnostics in snapshot, because getting code actions for fixes, requires access to the diagnostics and it only has access to a snapshot. Any ideas?

Reply:
Hm, yeah, it makes sense to add

```rust
struct DiagnosticsCollection {
    check_fixes: Arc<Vec<Fix>>
}
```

and copy `check_fixes` to the snapshots, while using `Arc::make_mut` to murate them on the main thread. This I think should avoid most coping.